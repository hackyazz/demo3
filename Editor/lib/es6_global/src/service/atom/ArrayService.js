

import * as Curry from "../../../../../node_modules/bs-platform/lib/es6/curry.js";
import * as Caml_obj from "../../../../../node_modules/bs-platform/lib/es6/caml_obj.js";
import * as Caml_option from "../../../../../node_modules/bs-platform/lib/es6/caml_option.js";
import * as Log$WonderLog from "../../../../../node_modules/wonder-log/lib/es6_global/src/Log.js";
import * as Contract$WonderLog from "../../../../../node_modules/wonder-log/lib/es6_global/src/Contract.js";
import * as Result$WonderEditor from "../../module/Result.js";
import * as OptionService$WonderEditor from "../primitive/OptionService.js";
import * as ArrayService$WonderCommonlib from "../../../../../node_modules/wonder-commonlib/lib/es6_global/src/ArrayService.js";
import * as StateEditorService$WonderEditor from "../state/editor/StateEditorService.js";
import * as MutableHashMapService$WonderCommonlib from "../../../../../node_modules/wonder-commonlib/lib/es6_global/src/MutableHashMapService.js";

function create(param) {
  return /* array */[];
}

function range(a, b) {
  var result = ArrayService$WonderCommonlib.createEmpty(/* () */0);
  for(var i = a; i <= b; ++i){
    result.push(i);
  }
  return result;
}

function unsafeGetFirst(arr) {
  return Contract$WonderLog.ensureCheck((function (r) {
                return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("first array element exist", "not"), (function (param) {
                              return Contract$WonderLog.assertNullableExist(r);
                            }));
              }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0), arr[0]);
}

function getFirst(arr) {
  return Caml_option.nullable_to_opt(arr[0]);
}

function unsafeGetLast(arr) {
  return Contract$WonderLog.ensureCheck((function (r) {
                return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("last array element exist", "not"), (function (param) {
                              return Contract$WonderLog.assertNullableExist(r);
                            }));
              }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0), arr[arr.length - 1 | 0]);
}

function getLast(arr) {
  return Caml_option.nullable_to_opt(arr[arr.length - 1 | 0]);
}

function removeLast(arr) {
  Contract$WonderLog.requireCheck((function (param) {
          return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("array length should >= 1", "not"), (function (param) {
                        return Contract$WonderLog.Operators[/* >= */7](arr.length, 1);
                      }));
        }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0));
  return OptionService$WonderEditor.unsafeGet(Caml_option.undefined_to_opt(arr.pop()));
}

function removeFirst(arr) {
  Contract$WonderLog.requireCheck((function (param) {
          return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("array length should >= 1", "not"), (function (param) {
                        return Contract$WonderLog.Operators[/* >= */7](arr.length, 1);
                      }));
        }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0));
  return OptionService$WonderEditor.unsafeGet(Caml_option.undefined_to_opt(arr.shift()));
}

function unsafeGetNth(index, arr) {
  return Contract$WonderLog.ensureCheck((function (r) {
                return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("array[" + (String(index) + "] element exist"), "not"), (function (param) {
                              return Contract$WonderLog.assertNullableExist(r);
                            }));
              }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0), arr[index]);
}

function hasItem(arr) {
  var match = arr.length > 0;
  if (match) {
    return true;
  } else {
    return false;
  }
}

function unshift(item, arr) {
  arr.unshift(item);
  return arr;
}

function push(item, arr) {
  arr.push(item);
  return arr;
}

function pushMany(itemArr, arr) {
  return ArrayService$WonderCommonlib.reduceOneParam((function (arr, item) {
                return push(item, arr);
              }), arr, itemArr);
}

function hasItemByFunc(func, arr) {
  return hasItem(arr.filter(Curry.__1(func)));
}

function removeDuplicateItems(buildKeyFunc, arr) {
  var resultArr = /* array */[];
  var map = MutableHashMapService$WonderCommonlib.createEmpty(/* () */0);
  for(var i = 0 ,i_finish = arr.length - 1 | 0; i <= i_finish; ++i){
    var item = arr[i];
    var key = buildKeyFunc(item);
    var match = MutableHashMapService$WonderCommonlib.get(key, map);
    if (match === undefined) {
      resultArr.push(item);
      MutableHashMapService$WonderCommonlib.set(key, item, map);
    }
    
  }
  return resultArr;
}

function exclude(excludeArr, arr) {
  return arr.filter((function (value) {
                return !excludeArr.includes(value);
              }));
}

function intersect(arr1, arr2) {
  return arr1.filter((function (value) {
                return arr2.includes(value);
              }));
}

function hasIntersect(arr1, arr2) {
  return intersect(arr1, arr2).length > 0;
}

function fastConcat(arr1, arr2) {
  return ArrayService$WonderCommonlib.reduceOneParam((function (arr1, value2) {
                return push(value2, arr1);
              }), arr1, arr2);
}

function fastConcatArrays(arrayArrays) {
  Contract$WonderLog.requireCheck((function (param) {
          return Contract$WonderLog.test(Log$WonderLog.buildAssertMessage("length >= 1", "not"), (function (param) {
                        return Contract$WonderLog.Operators[/* >= */7](arrayArrays.length, 1);
                      }));
        }), StateEditorService$WonderEditor.getStateIsDebug(/* () */0));
  return ArrayService$WonderCommonlib.reduceOneParam(fastConcat, arrayArrays[0], arrayArrays.slice(1));
}

function isEqual(arr1, arr2) {
  return Caml_obj.caml_equal(arr1.sort(), arr2.sort());
}

function isInclude(sourceArr, targetArr) {
  return targetArr.filter((function (value) {
                return !sourceArr.includes(value);
              })).length === 0;
}

function _addFailureFunc(param, param$1) {
  return /* tuple */[
          param[0],
          Curry._1(param[1], param$1[1])
        ];
}

function _handleFailAndSuceessFunc(param, s) {
  return /* tuple */[
          param[0],
          Curry._1(param[1], s)
        ];
}

function _handleSuceessFuncAndFailFunc(func, param) {
  return /* tuple */[
          param[0],
          Curry._1(func, param[1])
        ];
}

function traverseSameDataResultAndCollectByApply(traverseFunc, arr) {
  var func = Result$WonderEditor.SameDataResult[/* apply */2];
  var applyFunc = function (param) {
    var func$1 = Curry._1(func, param);
    return (function (param) {
        return Curry._4(func$1, param, _addFailureFunc, _handleFailAndSuceessFunc, _handleSuceessFuncAndFailFunc);
      });
  };
  var returnFunc = Result$WonderEditor.SameDataResult[/* success */0];
  return ArrayService$WonderCommonlib.reduceOneParam((function (resultArr, value) {
                return applyFunc(applyFunc(Curry._1(returnFunc, push))(Curry._1(traverseFunc, value)))(resultArr);
              }), Curry._1(returnFunc, /* array */[]), arr);
}

export {
  create ,
  range ,
  unsafeGetFirst ,
  getFirst ,
  unsafeGetLast ,
  getLast ,
  removeLast ,
  removeFirst ,
  unsafeGetNth ,
  hasItem ,
  unshift ,
  push ,
  pushMany ,
  hasItemByFunc ,
  removeDuplicateItems ,
  exclude ,
  intersect ,
  hasIntersect ,
  fastConcat ,
  fastConcatArrays ,
  isEqual ,
  isInclude ,
  _addFailureFunc ,
  _handleFailAndSuceessFunc ,
  _handleSuceessFuncAndFailFunc ,
  traverseSameDataResultAndCollectByApply ,
  
}
/* Log-WonderLog Not a pure module */
