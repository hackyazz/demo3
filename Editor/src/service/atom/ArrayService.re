let create = () => [||];

let range = (a: int, b: int) => {
  let result = WonderCommonlib.ArrayService.createEmpty();
  for (i in a to b) {
    Js.Array.push(i, result) |> ignore;
  };
  result;
};

let unsafeGetFirst = arr =>
  WonderCommonlib.ArrayService.unsafeGet(arr, 0)
  |> WonderLog.Contract.ensureCheck(
       r =>
         WonderLog.(
           Contract.(
             Operators.(
               test(
                 Log.buildAssertMessage(
                   ~expect={j|first array element exist|j},
                   ~actual={j|not|j},
                 ),
                 () =>
                 r |> assertNullableExist
               )
             )
           )
         ),
       StateEditorService.getStateIsDebug(),
     );

let getFirst = arr =>
  WonderCommonlib.ArrayService.unsafeGet(arr, 0) |> Js.toOption;

let unsafeGetLast = arr =>
  arr
  |> Js.Array.length
  |> (len => len - 1)
  |> WonderCommonlib.ArrayService.unsafeGet(arr)
  |> WonderLog.Contract.ensureCheck(
       r =>
         WonderLog.(
           Contract.(
             Operators.(
               test(
                 Log.buildAssertMessage(
                   ~expect={j|last array element exist|j},
                   ~actual={j|not|j},
                 ),
                 () =>
                 r |> assertNullableExist
               )
             )
           )
         ),
       StateEditorService.getStateIsDebug(),
     );

let getLast = arr =>
  arr
  |> Js.Array.length
  |> (len => len - 1)
  |> WonderCommonlib.ArrayService.unsafeGet(arr)
  |> Js.toOption;

let removeLast = arr => {
  WonderLog.Contract.requireCheck(
    () =>
      WonderLog.(
        Contract.(
          Operators.(
            test(
              Log.buildAssertMessage(
                ~expect={j|array length should >= 1|j},
                ~actual={j|not|j},
              ),
              () =>
              arr |> Js.Array.length >= 1
            )
          )
        )
      ),
    StateEditorService.getStateIsDebug(),
  );
  arr |> Js.Array.pop |> OptionService.unsafeGet;
};

let removeFirst = arr => {
  WonderLog.Contract.requireCheck(
    () =>
      WonderLog.(
        Contract.(
          Operators.(
            test(
              Log.buildAssertMessage(
                ~expect={j|array length should >= 1|j},
                ~actual={j|not|j},
              ),
              () =>
              arr |> Js.Array.length >= 1
            )
          )
        )
      ),
    StateEditorService.getStateIsDebug(),
  );
  arr |> Js.Array.shift |> OptionService.unsafeGet;
};

let unsafeGetNth = (index, arr) =>
  WonderCommonlib.ArrayService.unsafeGet(arr, index)
  |> WonderLog.Contract.ensureCheck(
       r =>
         WonderLog.(
           Contract.(
             Operators.(
               test(
                 Log.buildAssertMessage(
                   ~expect={j|array[$index] element exist|j},
                   ~actual={j|not|j},
                 ),
                 () =>
                 r |> assertNullableExist
               )
             )
           )
         ),
       StateEditorService.getStateIsDebug(),
     );

let hasItem = arr => arr |> Js.Array.length > 0;

let unshift = (item, arr) => {
  arr |> Js.Array.unshift(item) |> ignore;
  arr;
};

let push = (item, arr) => {
  arr |> Js.Array.push(item) |> ignore;
  arr;
};

let pushMany = (itemArr, arr) =>
  itemArr
  |> WonderCommonlib.ArrayService.reduceOneParam(
       (. arr, item) => arr |> push(item),
       arr,
     );

let hasItemByFunc = (func, arr) => arr |> Js.Array.filter(func) |> hasItem;

let removeDuplicateItems = (buildKeyFunc, arr) => {
  open WonderCommonlib;

  let resultArr = [||];
  let map = MutableHashMapService.createEmpty();

  for (i in 0 to Js.Array.length(arr) - 1) {
    let item = Array.unsafe_get(arr, i);
    let key = buildKeyFunc(. item);

    switch (MutableHashMapService.get(key, map)) {
    | None =>
      Js.Array.push(item, resultArr) |> ignore;
      MutableHashMapService.set(key, item, map) |> ignore;
    | Some(_) => ()
    };
  };

  resultArr;
};

let exclude = (excludeArr, arr) =>
  arr |> Js.Array.filter(value => ! (excludeArr |> Js.Array.includes(value)));

let intersect = (arr1, arr2) =>
  arr1 |> Js.Array.filter(value => arr2 |> Js.Array.includes(value));

let hasIntersect = (arr1, arr2) =>
  intersect(arr1, arr2) |> Js.Array.length > 0;

let fastConcat = (arr1, arr2) =>
  arr2
  |> WonderCommonlib.ArrayService.reduceOneParam(
       (. arr1, value2) => arr1 |> push(value2),
       arr1,
     );

let fastConcatArrays = arrayArrays => {
  WonderLog.Contract.requireCheck(
    () =>
      WonderLog.(
        Contract.(
          Operators.(
            test(
              Log.buildAssertMessage(
                ~expect={j|length >= 1|j},
                ~actual={j|not|j},
              ),
              () =>
              Js.Array.length(arrayArrays) >= 1
            )
          )
        )
      ),
    StateEditorService.getStateIsDebug(),
  );

  arrayArrays
  |> Js.Array.sliceFrom(1)
  |> WonderCommonlib.ArrayService.reduceOneParam(
       (. arr1, arr) => fastConcat(arr1, arr),
       Array.unsafe_get(arrayArrays, 0),
     );
};

let isEqual = (arr1, arr2) =>
  arr1 |> Js.Array.sortInPlace == (arr2 |> Js.Array.sortInPlace);

let isInclude = (sourceArr, targetArr) =>
  targetArr
  |> Js.Array.filter(value => ! (sourceArr |> Js.Array.includes(value)))
  |> Js.Array.length === 0;

let _addFailureFunc = ((msg1, value1), (msg2, value2)) => (
  msg1,
  value1(value2),
);

let _handleFailAndSuceessFunc = ((msg, value1), s) => (msg, value1(s));

let _handleSuceessFuncAndFailFunc = (func, (msg2, value2)) => (
  msg2,
  func(value2),
);

let traverseSameDataResultAndCollectByApply = (traverseFunc, arr) => {
  let applyFunc =
    Result.SameDataResult.apply(
      ~addFailureFunc=_addFailureFunc,
      ~handleFailAndSuceessFunc=_handleFailAndSuceessFunc,
      ~handleSuceessFuncAndFailFunc=_handleSuceessFuncAndFailFunc,
    );
  let returnFunc = Result.SameDataResult.success;

  arr
  |> WonderCommonlib.ArrayService.reduceOneParam(
       (. resultArr, value) =>
         applyFunc(
           ~switchFunc=returnFunc(push),
           ~result=traverseFunc(value),
         )
         |> applyFunc(~switchFunc=_, ~result=resultArr),
       returnFunc([||]),
     );
};