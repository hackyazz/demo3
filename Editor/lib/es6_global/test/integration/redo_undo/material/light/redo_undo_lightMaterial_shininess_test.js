

import * as Curry from "../../../../../../../node_modules/bs-platform/lib/es6/curry.js";
import * as Sinon from "../../../../../../../node_modules/wonder-bs-sinon/lib/es6_global/src/sinon.js";
import * as Sinon$1 from "sinon";
import * as ArrayService$WonderEditor from "../../../../../src/service/atom/ArrayService.js";
import * as RedoUndoTool$WonderEditor from "../../tool/RedoUndoTool.js";
import * as GameObjectTool$WonderEditor from "../../../../tool/GameObjectTool.js";
import * as EventListenerTool$WonderEditor from "../../../../unit/tool/EventListenerTool.js";
import * as MainEditorAssetTool$WonderEditor from "../../../asset/tool/MainEditorAssetTool.js";
import * as MainEditorSceneTool$WonderEditor from "../../../../tool/MainEditorSceneTool.js";
import * as BuildComponentForCurryTool$WonderEditor from "../../../../tool/BuildComponentForCurryTool.js";
import * as MainEditorLightMaterialTool$WonderEditor from "../../../inspector/composable_component/sceneTree_inspector/renderGroup/material/tool/MainEditorLightMaterialTool.js";

describe("redo_undo: lightMaterial shininess", (function () {
        var sandbox = Sinon.getSandboxDefaultVal(/* () */0);
        var _changeShininess = function (value) {
          GameObjectTool$WonderEditor.getCurrentGameObjectLightMaterial(/* () */0);
          MainEditorLightMaterialTool$WonderEditor.changeShininess(undefined, value, /* () */0);
          return MainEditorLightMaterialTool$WonderEditor.blurShininess(undefined, undefined, undefined, value, /* () */0);
        };
        var _simulateTwiceChangeShininess = function () {
          _changeShininess(1.1);
          return _changeShininess(12.12);
        };
        var _beforeEach = function () {
          return MainEditorSceneTool$WonderEditor.createDefaultScene(sandbox, (function () {
                        MainEditorAssetTool$WonderEditor.initAssetTree(/* () */0);
                        return MainEditorSceneTool$WonderEditor.setFirstCubeToBeCurrentSceneTreeNode(/* () */0);
                      }));
        };
        beforeEach((function () {
                sandbox[0] = Sinon$1.sandbox.create();
                MainEditorSceneTool$WonderEditor.initState(sandbox, undefined, undefined, /* () */0);
                return Curry._1(EventListenerTool$WonderEditor.stubGetElementByIdReturnFakeDom, Curry._1(EventListenerTool$WonderEditor.buildFakeDom, /* () */0));
              }));
        afterEach((function () {
                return Curry._1(Sinon.restoreSandbox, sandbox[0]);
              }));
        return RedoUndoTool$WonderEditor.testRedoUndoTwoStep(sandbox, "prepare first step: set currentSceneTreeNode", /* tuple */[
                    _simulateTwiceChangeShininess,
                    _beforeEach,
                    (function () {
                        return /* () */0;
                      })
                  ], BuildComponentForCurryTool$WonderEditor.buildLightMaterial);
      }));

export {
  
}
/*  Not a pure module */
