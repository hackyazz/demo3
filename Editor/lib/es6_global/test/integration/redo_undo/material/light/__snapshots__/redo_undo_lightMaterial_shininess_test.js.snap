// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: lightMaterial shininess prepare first step: set currentSceneTreeNode test redo operate test redo one step undo step which from second to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-light-material"
>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffcuse color
    </div>
    <div
      className="item-content item-color"
    >
      <div
        className="color-hex"
        onClick={[Function]}
        style={
          Object {
            "background": "#ffffff",
          }
        }
      />
      <div
        className="color-select"
        onClick={[Function]}
      >
        <img
          src="./public/img/color.png"
        />
      </div>
    </div>
  </article>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffuse map
    </div>
    <div
      className="item-content item-texture"
    >
      <div
        className="texture-img"
        onClick={[Function]}
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          className="img-dragBg"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
      </div>
      <button
        className="texture-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </article>
  <article
    className="inspector-item wonder-float-input"
  >
    <div
      className="item-header component-label"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
    >
      Shininess
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="12.12"
      />
    </div>
  </article>
</article>
`;

exports[`redo_undo: lightMaterial shininess prepare first step: set currentSceneTreeNode test redo operate test redo two step undo step which from second to zero,redo step which from zero to second 1`] = `
<article
  className="wonder-light-material"
>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffcuse color
    </div>
    <div
      className="item-content item-color"
    >
      <div
        className="color-hex"
        onClick={[Function]}
        style={
          Object {
            "background": "#ffffff",
          }
        }
      />
      <div
        className="color-select"
        onClick={[Function]}
      >
        <img
          src="./public/img/color.png"
        />
      </div>
    </div>
  </article>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffuse map
    </div>
    <div
      className="item-content item-texture"
    >
      <div
        className="texture-img"
        onClick={[Function]}
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          className="img-dragBg"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
      </div>
      <button
        className="texture-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </article>
  <article
    className="inspector-item wonder-float-input"
  >
    <div
      className="item-header component-label"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
    >
      Shininess
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="12.12"
      />
    </div>
  </article>
</article>
`;

exports[`redo_undo: lightMaterial shininess prepare first step: set currentSceneTreeNode test undo operate test not undo 1`] = `
<article
  className="wonder-light-material"
>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffcuse color
    </div>
    <div
      className="item-content item-color"
    >
      <div
        className="color-hex"
        onClick={[Function]}
        style={
          Object {
            "background": "#ffffff",
          }
        }
      />
      <div
        className="color-select"
        onClick={[Function]}
      >
        <img
          src="./public/img/color.png"
        />
      </div>
    </div>
  </article>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffuse map
    </div>
    <div
      className="item-content item-texture"
    >
      <div
        className="texture-img"
        onClick={[Function]}
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          className="img-dragBg"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
      </div>
      <button
        className="texture-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </article>
  <article
    className="inspector-item wonder-float-input"
  >
    <div
      className="item-header component-label"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
    >
      Shininess
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="12.12"
      />
    </div>
  </article>
</article>
`;

exports[`redo_undo: lightMaterial shininess prepare first step: set currentSceneTreeNode test undo operate test undo one step step which from second to first 1`] = `
<article
  className="wonder-light-material"
>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffcuse color
    </div>
    <div
      className="item-content item-color"
    >
      <div
        className="color-hex"
        onClick={[Function]}
        style={
          Object {
            "background": "#ffffff",
          }
        }
      />
      <div
        className="color-select"
        onClick={[Function]}
      >
        <img
          src="./public/img/color.png"
        />
      </div>
    </div>
  </article>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffuse map
    </div>
    <div
      className="item-content item-texture"
    >
      <div
        className="texture-img"
        onClick={[Function]}
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          className="img-dragBg"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
      </div>
      <button
        className="texture-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </article>
  <article
    className="inspector-item wonder-float-input"
  >
    <div
      className="item-header component-label"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
    >
      Shininess
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="12.12"
      />
    </div>
  </article>
</article>
`;

exports[`redo_undo: lightMaterial shininess prepare first step: set currentSceneTreeNode test undo operate test undo two step step which from second to zero 1`] = `
<article
  className="wonder-light-material"
>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffcuse color
    </div>
    <div
      className="item-content item-color"
    >
      <div
        className="color-hex"
        onClick={[Function]}
        style={
          Object {
            "background": "#ffffff",
          }
        }
      />
      <div
        className="color-select"
        onClick={[Function]}
      >
        <img
          src="./public/img/color.png"
        />
      </div>
    </div>
  </article>
  <article
    className="inspector-item"
  >
    <div
      className="item-header"
    >
      Diffuse map
    </div>
    <div
      className="item-content item-texture"
    >
      <div
        className="texture-img"
        onClick={[Function]}
        style={
          Object {
            "opacity": "1",
          }
        }
      >
        <div
          className="img-dragBg"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
      </div>
      <button
        className="texture-remove"
        onClick={[Function]}
      >
        Remove
      </button>
    </div>
  </article>
  <article
    className="inspector-item wonder-float-input"
  >
    <div
      className="item-header component-label"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
    >
      Shininess
    </div>
    <div
      className="item-content"
    >
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="1.1"
      />
    </div>
  </article>
</article>
`;
